sensor/kubernetes/listener/resources/networkpolicy.go

```diff
@@ -18,34 +20,80 @@ type networkPolicyStore interface {
+// Process processes a network policy resource event, and returns the sensor events to generate.
+func (h *networkPolicyDispatcher) ProcessEvent(obj, _ interface{}, action central.ResourceAction) []*central.SensorEvent {
+	np := obj.(*networkingV1.NetworkPolicy)
 
-	switch action {
-	case central.ResourceAction_CREATE_RESOURCE, central.ResourceAction_UPDATE_RESOURCE:
-		h.store.Upsert(netPolicy)
-	case central.ResourceAction_REMOVE_RESOURCE:
-		h.store.Delete(netPolicy.GetId(), netPolicy.GetNamespace())
+	roxNetpol := networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: np}.ToRoxNetworkPolicy()
+
+	if features.NetworkPolicySystemPolicy.Enabled() {
+		sel, isEmpty := h.getSelector(roxNetpol, action)
+		if action == central.ResourceAction_REMOVE_RESOURCE {
+			h.netpolStore.Delete(roxNetpol.GetId(), roxNetpol.GetNamespace())
+		} else {
+			h.netpolStore.Upsert(roxNetpol)
+		}
+
+		h.updateDeploymentsFromStore(roxNetpol, sel, isEmpty)
 	}
 
 	return []*central.SensorEvent{
 		{
 			Id:     string(np.UID),
 			Action: action,
 			Resource: &central.SensorEvent_NetworkPolicy{
-				NetworkPolicy: networkPolicyConversion.KubernetesNetworkPolicyWrap{NetworkPolicy: np}.ToRoxNetworkPolicy(),
+				NetworkPolicy: roxNetpol,
 			},
 		},
 	}
 }
+
+func (h *networkPolicyDispatcher) getSelector(np *storage.NetworkPolicy, action central.ResourceAction) (selector, bool) {
+	var sel selector
+	isEmpty := true
+	oldWrap := h.netpolStore.Get(np.GetId())
+	if oldWrap != nil {
+		sel = SelectorFromMap(oldWrap.GetSpec().GetPodSelector().GetMatchLabels())
+		isEmpty = len(oldWrap.GetSpec().GetPodSelector().GetMatchLabels()) == 0
+	}
+
+	if action == central.ResourceAction_UPDATE_RESOURCE {
+		if sel != nil {
+			sel = or(sel, SelectorFromMap(np.GetSpec().GetPodSelector().GetMatchLabels()))
+			isEmpty = isEmpty && len(np.GetSpec().GetPodSelector().GetMatchLabels()) == 0
+		} else {
+			sel = SelectorFromMap(np.GetSpec().GetPodSelector().GetMatchLabels())
+			isEmpty = len(np.GetSpec().GetPodSelector().GetMatchLabels()) == 0
+		}
+	} else if action == central.ResourceAction_CREATE_RESOURCE {
+		sel = SelectorFromMap(np.GetSpec().GetPodSelector().GetMatchLabels())
+		isEmpty = len(np.GetSpec().GetPodSelector().GetMatchLabels()) == 0
+	}
+	return sel, isEmpty
+}
+
+func (h *networkPolicyDispatcher) updateDeploymentsFromStore(np *storage.NetworkPolicy, sel selector, isEmptySelector bool) {
+	if !isEmptySelector {
+		for _, deploymentWrap := range h.deploymentStore.getMatchingDeployments(np.GetNamespace(), sel) {
+			h.detector.ProcessDeployment(deploymentWrap.GetDeployment(), central.ResourceAction_UPDATE_RESOURCE)
+		}
+	} else {
+		// Network Policies with no selector match with all the deployments in the namespace
+		for _, deploymentWrap := range h.deploymentStore.getAllDeploymentsInNamespace(np.GetNamespace()) {
+			h.detector.ProcessDeployment(deploymentWrap.GetDeployment(), central.ResourceAction_UPDATE_RESOURCE)
+		}
+	}
+}
```

Comment:
1. At first I thought of using [Empty()](https://github.com/kubernetes/apimachinery/blob/6550efdb74445c227aa0bd74017b90a9a8cd07e2/pkg/labels/selector.go#L54). The problem; it returns `false` when the selector is `nothingSelector` which we use [here](https://github.com/stackrox/stackrox/blob/aad0c634907b667353acb6a31db1a3b34ad514b9/sensor/kubernetes/listener/resources/selector.go#L33). Also using the `len` of the selector does not work because if we are in the case of an old network policy with no selector and we update the `PodSelector`, we will have `selectorDisjunction` like `[{}, {"key1: "value1", "key2": "value2"}] and this should be considered as empty.
2. Right now, if the map is empty we create a `nothingSelector` which returns always `false` so [here](https://github.com/stackrox/stackrox/blob/9575e3acc469acfac27b290f5d92c4bb7b473e0f/sensor/kubernetes/listener/resources/deployment_store.go#L83-L85) we will never match anything when we should match everything.

Reply:
I see. So involving `nothingSelector` makes it clear that empty selector matches everything and nothing selector matches nothing. I am thinking how to make it clearer when reading the code 🤔  Maybe we could rename `isEmptySelector` into `matchesAll` or `matchAll`?

Or maybe split it into two separate functions - one for all deployments `functionName(np *storage.NetworkPolicy)` and the other `function2(np *storage.NetworkPolicy, sel selector)`?

Reply:
I like the renaming it to `matchesAll` approach. I wouldn't separate the logic in two functions because in the future we might need to add some logic in loops (i.e if we store the augmented object in the deployment wrap) and I think is better if we don't have duplicated code. At that point I would probably join both loops into one.
```go
var deployments []*deploymentWrap
if matchesAll {
    deployments = h.deploymentStore.getAllDeploymentsInNamespace(np.GetNamespace())
} else {
    deployments = h.deploymentStore.getMatchingDeployments(np.GetNamespace(), sel)
}
for _, d := range deployments {
...
}
```

What do you think?

Reply:
Yes, sounds good - this should be enough to get what is happening here. You may optionally add a comment explaining why we use it (and mention `nothingSelector`) - up to you.

Reply:
Done